mailpoet/lib/Cron/Workers/BackfillEngagementData.php

```diff
@@ -0,0 +1,48 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Cron\Workers;
+
+use MailPoet\Entities\ScheduledTaskEntity;
+use MailPoet\Subscribers\EngagementDataBackfiller;
+use MailPoet\WP\Functions as WPFunctions;
+
+class BackfillEngagementData extends SimpleWorker {
+  const TASK_TYPE = 'backfill_engagement_data';
+  const BATCH_SIZE = 100;
+  const AUTOMATIC_SCHEDULING = false;
+  const SUPPORT_MULTIPLE_INSTANCES = false;
+
+  /** @var EngagementDataBackfiller */
+  private $engagementDataBackfiller;
+
+  public function __construct(
+    EngagementDataBackfiller $engagementDataBackfiller,
+    WPFunctions $wp
+  ) {
+    parent::__construct($wp);
+    $this->engagementDataBackfiller = $engagementDataBackfiller;
+  }
+
+  public function processTaskStrategy(ScheduledTaskEntity $task, $timer) {
+    $meta = $task->getMeta();
+
+    $lastSubscriberId = $meta['last_subscriber_id'] ?? 0;
+
+    do {
+      $this->cronHelper->enforceExecutionLimit($timer);
+      $batch = $this->engagementDataBackfiller->getBatch($lastSubscriberId);
+      if (empty($batch)) {
+        break;
+      }
+      $this->engagementDataBackfiller->updateBatch($batch);
+      $lastSubscriberId = $this->engagementDataBackfiller->getLastProcessedSubscriberId();
+    } while (count($batch) === self::BATCH_SIZE);
+
+    $meta['last_subscriber_id'] = $this->engagementDataBackfiller->getLastProcessedSubscriberId();
+    $task->setMeta($meta);
+    $this->scheduledTasksRepository->persist($task);
+    $this->scheduledTasksRepository->flush();
```

Comment:
Should we reschedule even when the backfill has been completed? To me it seems like we would restart with subscriber ID `0` when all have been processed:

The `emty($batch)` breaks the `do {} while` and then `getLastProcessedSubscriberId` will return 0 which will be stored as meta value and we start again.

Reply:
You're right that we could end up saving the last processed ID as `0`, but I don't think that would cause the task to restart, and I think it might be a bit of an edge case.

The `empty` check would indeed break the `do` loop, but `getLastProcessedSubscriberId` would only return `0` if this happens during the first batch it attempts to process during a cron run. If it processed a subscriber from the previous batch, it would still be set in the backfiller's state.

I think what would need to happen is the execution limit would have to be hit immediately after processing the last batch. The next time the task runs, the batch would be empty and `getLastProcessedSubscriberId` would return the default of `0`.

I updated the code in b3a47865e6cabca0c3499cee29727a2b3761f2c7 to account for this case.

As for restarting, I don't think I'm doing any scheduling based off the `last_subscriber_id` in the meta. As long as the `processTaskStrategy` code completes before the execution limit, the task should get marked as complete. If it exceeds the limit, then I think it will [resume during the next run ](https://github.com/mailpoet/mailpoet/blob/MAILPOET-5404%2Fbackfill_engagement_data/mailpoet/lib/Cron/CronHelper.php#L227-L227).

I could definitely be wrong!

Reply:
You can disregard a lot of my previous comment. I realized I needed to save the meta value during the loop anyway, so I simplified everything in a248d485d381eb3b18b3f69734b478f0786a9943.